### PR TITLE
feat(auth): Native Android bridge

### PR DESCRIPTION
**Pull request by dnys1**
Adds the Android implementation for the two-way Auth bridge allowing native Android plugins like API/DataStore to call into Flutter for retrieving auth credentials. Further a bridge is established to allow the Dart plugin to call the native addPlugin method with its configuration.

commit-id:08874ee6

---

**Stack**:
- #1655
- #1654
- #1652
- #1648
- #1647
- #1645
- #1642 ⬅
- #1639

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

**Review by Jordan-Nelson — APPROVED** [review]

File: packages/amplify/amplify_flutter_android/android/build.gradle [hunk 1]
```diff
@@ -72,15 +72,15 @@ android {
 dependencies {
     api amplifyCore
 
-    implementation 'com.amplifyframework:core:1.33.0'
-    implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-android:1.3.9'
+    implementation 'com.amplifyframework:core:1.36.0-dev-preview.0'
```
Comment (haverchuck) [inline packages/amplify/amplify_flutter_android/android/build.gradle:75]: What is the purpose of using the Android dev-preview?

File: packages/amplify/amplify_flutter/example/android/app/build.gradle [hunk 2]
```diff
@@ -40,7 +40,7 @@ android {
     defaultConfig {
         // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
         applicationId "com.amazonaws.amplify.amplify_flutter_example"
-        minSdkVersion 21
+        minSdkVersion 24
```
Comment (haverchuck) [inline packages/amplify/amplify_flutter/example/android/app/build.gradle:43]: Are we going to merging all of this migration code into a separate dev-preview branch, or migrating everything into main and then keeping the dev-preview on a seperate branch where the new packages are enabled view pubspecs?

File: packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/types/FlutterConfirmResetPasswordRequest.kt [hunk 3]
```diff
@@ -1,52 +0,0 @@
...
- * permissions and limitations under the License.
- */
-
-package com.amazonaws.amplify.amplify_auth_cognito.types
-
-import com.amazonaws.amplify.amplify_core.exception.ExceptionMessages
-import com.amazonaws.amplify.amplify_core.exception.InvalidRequestException
-import com.amplifyframework.auth.cognito.options.AWSCognitoAuthConfirmResetPasswordOptions
-
-data class FlutterConfirmResetPasswordRequest(val map: HashMap<String, *>) {
```
Comment (haverchuck) [inline packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/types/FlutterConfirmResetPasswordRequest.kt:22]: Good riddance to these classes!

File: packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/NativeAuthPlugin.kt [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+                    AuthSessionResult.success(session.userSub),
+                    userPoolTokens,
+                )
+                onSuccess.accept(authSession)
+            }
+        }
+    }
+
+    override fun getCurrentUser(): AuthUser {
+        fail("getCurrentUser is not supported")
```
Comment (haverchuck) [inline packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/NativeAuthPlugin.kt:130]: nit: Maybe we should use a more helpful message, perhaps conveying why these methods aren't supported (not a huge deal either way though).

File: packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/NativeAuthPlugin.kt [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+import com.amplifyframework.auth.result.AuthSignUpResult
+import com.amplifyframework.auth.result.AuthUpdateAttributeResult
+import com.amplifyframework.core.Action
+import com.amplifyframework.core.Consumer
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.launch
+import org.json.JSONObject
+
+class NativeAuthPlugin(
```
Comment (haverchuck) [inline packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/NativeAuthPlugin.kt:61]: I think we should probably add some documentation here conveying the purpose of this plugin (for external contributers and such).

File: packages/amplify/amplify_flutter_android/android/build.gradle (hunk 1 above)
Comment (dnys1) [reply packages/amplify/amplify_flutter_android/android/build.gradle:75]: Answered offline. We'll be using native dev-previews for our dev-preview, and likewise for GA.

File: packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/NativeAuthPlugin.kt (hunk 4 above)
Comment (dnys1) [reply packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/NativeAuthPlugin.kt]: Done 👍

File: packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/NativeAuthPlugin.kt (hunk 5 above)
Comment (dnys1) [reply packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/NativeAuthPlugin.kt:64]: Done 👍

**Review by fjnoyp — APPROVED** [review]
Looks good, thanks for getting to this!

**Review by haverchuck — APPROVED** [review]

**Pull request merged**